docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md

```diff
@@ -1,3 +1,570 @@
+
 # Writing a ClusterClass
 
-**Under Construction**
+While a ClusterClass can be also used for a single Cluster, usually a ClusterClass is used across many
+clusters. Therefore a ClusterClass becomes more useful when it's flexible enough to support a wide
+range of use cases. The goal of this document is to explain how ClusterClasses can be written in
+general and how they can be written in a way that they are flexible enough to support many use cases.
+
+**Table of Contents**
+
+* [Basic ClusterClass](#basic-clusterclass)
+* [ClusterClass with MachineHealthChecks](#clusterclass-with-machinehealthchecks)
+* [ClusterClass with patches](#clusterclass-with-patches)
+* [Advanced features of ClusterClass with patches](#advanced-features-of-clusterclass-with-patches)
+    * [Builtin variables](#builtin-variables)
+    * [JSON patch values](#json-patch-values)
+    * [Complex variables](#complex-variables)
+    * [Optional patches](#optional-patches)
+    * [MachineDeployment variable overrides](#machinedeployment-variable-overrides)
+
+## Basic ClusterClass
+
+The following example shows a basic ClusterClass. It contains templates to shape the control plane, 
+infrastructure and workers of a Cluster.  When a Cluster is using this ClusterClass, the templates 
+are used to generate the objects of the managed topology of the Cluster.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: docker-clusterclass-v0.1.0
+spec:
+  controlPlane:
+    ref:
+      apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+      kind: KubeadmControlPlaneTemplate
+      name: docker-clusterclass-v0.1.0
+      namespace: default
+    machineInfrastructure:
+      ref:
+        kind: DockerMachineTemplate
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        name: docker-clusterclass-v0.1.0
+        namespace: default
+  infrastructure:
+    ref:
+      apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+      kind: DockerClusterTemplate
+      name: docker-clusterclass-v0.1.0-control-plane
+      namespace: default
+  workers:
+    machineDeployments:
+    - class: default-worker
+      template:
+        bootstrap:
+          ref:
+            apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+            kind: KubeadmConfigTemplate
+            name: docker-clusterclass-v0.1.0-default-worker
+            namespace: default
+        infrastructure:
+          ref:
+            apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+            kind: DockerMachineTemplate
+            name: docker-clusterclass-v0.1.0-default-worker
+            namespace: default
+```
+
+The following example shows a Cluster using this ClusterClass. In this case a `KubeadmControlPlane` 
```

Comment:
Might be easier for readers to follow along if there are  example templates here so users can follow this guide to create and edit a working ClusterClass and Cluster using the yamls here.

Reply:
we can eventually link to the full yaml, but I see value also in keeping thing simple and short

Reply:
My idea was to essentially only make it a series of snippets to show the relevant concepts.
Otherwise it would be so much YAML to dig through and we would have to keep maintaining them.

Reply:
What do we think about referencing the CAPD ClusterClass cluster-template that we are also publishing here?

Reply:
@sbueringer  FYI @ykakarap have found a way to provide collapsible sections with the full yaml, we can probably use the same approach here as well

Reply:
At the moment I've added a link to the published YAMLs. I'm not sure if we want to have too many copies of CAPD ClusterClasses and Clusters, especially as probably nobody tested them and it's hard to ensure manually the full examples keep working.